scripts/openwebnet-discovery.nse

```diff
@@ -0,0 +1,199 @@
+local string = require "string"
+
+description = [[
+OpenWebNet is a communications protocol developed by Bticino since 2000.
+Retrieves the Gateway and device type. Retrieves the count and addresses
+of lights, multimedia and many other services running on server/servers.
+]]
+
+---
+-- @usage
+-- nmap --script openwebnet-discovery
+--
+-- @output
+-- | openwebnet-discovery:
+-- |   Gateway: *#*1##*#13**15*12##*#1##
+-- |   Heating: 1
+-- |   Power Management: 1
+-- |   Multimedia: 1
+-- |   Device: F453AV
+-- |   Automation: 3
+-- |   Door Entry System: 2
+-- |   Burglar Alarm: 1
+-- |_  Lighting: 114
+--
+-- Version: 0.1, Updated on 21/06/2017
+
+author = "Rewanth Cool"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery", "safe"}
+
+portrule = shortport.port_or_service(20000, "openwebnet")
+
+local device = {}
+device[2] = "MHServer"
+device[4] = "MH200"
+device[6] = "F452"
+device[7] = "F452V"
+device[11] = "MHServer2"
+device[12] = "F453AV"
+device[13] = "H4684"
+device[15] = "F427 (Gateway Open-KNX)"
+device[16] = "F453"
+device[23] = "H4684"
+device[27] = "L4686SDK"
+device[44] = "MH200N"
+device[51] = "F454"
+device[200] = "F454 (new?)"
+
+local who = {}
+who[0] = "Scenarios"
+who[1] = "Lighting"
+who[2] = "Automation"
+who[3] = "Power Management"
+who[4] = "Heating"
+who[5] = "Burglar Alarm"
+who[6] = "Door Entry System"
+who[7] = "Multimedia"
+who[9] = "Auxiliary"
+who[13] = "Device Communication"
+who[14] = "Light+shutters actuators lock"
+who[15] = "WHO=15 - CEN"
+who[16] = "Sound System"
+who[17] = "Scenario Programming"
+who[18] = "Energy Management"
+who[24] = "Lighting Management"
+who[25] = "WHO=25 - CEN plus"
+who[1000] = "Diagnostic"
+who[1001] = "Automation Diagnostic"
+who[1004] = "Heating Diagnostic"
+who[1008] = "Door Entry System Diagnostic"
+who[1013] = "Device Diagnostic"
+
+-- Initiates a socket connection and returns the received data.
+local function get_socket(host, port, request)
+
+  local sd, response, early_resp = comm.opencon(host, port, request)
+
+  if sd == nil then
+    return nil, nil, "Socket connection error."
+  end
+
+  if not response then
+    return nil, nil, "Poor internet connection or no response."
+  end
+
+  if response == "*#*0##" then
+    return nil, nil, "Received a negative ACK as response."
+  end
+
+  stdnse.sleep(2)
+
+  -- Request for fetching Gateway address
+  sd:send(request)
+
+  local status, data = sd:receive_buf("*#*1##", false)
+  if data == "EOF" then
+    return nil, nil, "Received EOF with no response."
+  end
+
+  if data == "TIMEOUT" then
+    return nil, nil, "Timeout occurred."
+  end
+
+  -- Ignore if the response is EOF which is of length 3
+  -- Gateway length will be greater than 3
+  if #data > 3 then
+    -- Appending the delimiters to the data received for showing output
+    data = "*#*1##" .. data .. "1##"
+    return sd, data, nil
+  end
+end
+
+-- Removes *#*1## from the beginning and ending
```

Comment:
Let's take the request for the device's own IP address as an example, since [the reference page](http://www.pimyhome.org/wiki/index.php/WHO%3D13_-_Device_Communication) says that Gateway IP is only sometimes supported. The request is `*#13**10##` and the response is:

```
*#13**10*192*168*1*40##*#*1##
|||  ||  |           | |_ ACK
|||  ||  |           |_ End of message
|||  ||  |_ VALUE: 192.168.1.40
|||  ||_ DIMENSION: IP Address
|||  |_ WHERE: empty 
|||_ WHO: Device Communication
||_ this is a Dimension message
|_ Start of message
```

So for any request, we should parse the response into a series of messages (usually there will be only 1 plus an ACK), validate that it corresponds to the request we sent (WHO, WHERE, and DIMENSION match), and then convert the VALUE as appropriate. In this case, it's clear that the value is a list of the octets of an IP address. We could probably just replace the `*` with `.` in that case. Time and Date are also interesting, so we'll have to extract each value and [convert it into a readable string](https://nmap.org/nsedoc/lib/stdnse.html#format_timestamp). When we find an ACK or NACK message, that means we're done parsing; NACK means the command was bad or not supported.